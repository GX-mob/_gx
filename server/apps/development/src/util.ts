import chalk from "chalk";

export function log(title?: string, content?: string, formatTitle = true) {
  if (!title && !content) {
    return (title: string) => (content: string) =>
      log(title, content, typeof title === "undefined");
  }

  if (!content) {
    return console.log(title);
  }

  console.log(formatTitle ? chalk.bold.inverse(` ${title} `) : title, content);
}
