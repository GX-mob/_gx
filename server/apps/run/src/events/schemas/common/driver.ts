import { UserBasic } from "./user-basic";
import { Position } from "../position";
import { Configuration } from "../configuration";

export type Driver = UserBasic & {
  position: Position;
  /**
   * Driver state list:
   * * undefined | 1 = Idle
   * * 2 = Searching
   * * 3 = Running
   */
  state?: 1 | 2 | 3;
  /**
   * Driver ride match configuration
   */
  config: Configuration;
};
