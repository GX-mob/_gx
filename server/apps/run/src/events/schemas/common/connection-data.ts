import { Configuration } from "../configuration";
import { UserBasic } from "./user-basic";
import { Driver } from "./driver";
import { Ride, USERS_ROLES } from "@app/repositories";

export type ConnectionData = UserBasic & {
  mode: USERS_ROLES;
  /**
   * Sockets that observe some events of this socket
   */
  observers: { socketId: string; p2p: boolean }[];
  config?: Configuration;
  /**
   * Current rides
   * * Voyager only
   */
  rides?: Ride["pid"][];
  /**
   * Driver state
   * * Driver only
   */
  state?: Driver["state"];
};
