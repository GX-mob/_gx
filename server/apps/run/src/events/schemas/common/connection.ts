import { Configuration } from "../configuration";
import { UserBasic } from "./user-basic";
import { Driver } from "./driver";
import { Models } from "@gx-mob/http-service";
import { USERS_ROLES } from "@app/database";

export type Connection = UserBasic & {
  /**
   * Usage mode
   * * 1 = Voyager
   * * 2 = Driver
   */
  mode: USERS_ROLES;
  /**
   * Sockets that observe some events of this socket
   */
  observers: { socketId: string; p2p: boolean }[];
  config?: Configuration;
  rides?: Models.Ride[];
  state?: Driver["state"];
};
