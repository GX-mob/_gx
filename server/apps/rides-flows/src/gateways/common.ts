import { ForbiddenException } from "@nestjs/common";
import {
  MessageBody,
  SubscribeMessage,
  OnGatewayInit,
  OnGatewayConnection,
  ConnectedSocket,
} from "@nestjs/websockets";
import { PinoLogger } from "nestjs-pino";
import { ConfigService } from "@nestjs/config";
import { Server, Socket } from "socket.io";
import { auth, storageAdapters, unique } from "extensor";
import { SocketService } from "@app/socket";
import { AuthService } from "@app/auth";
import { CacheService } from "@app/cache";
import { retryUnderHood } from "@app/helpers/util";
import { IUser, IRide, UserRoles } from "@shared/interfaces";
import {
  PendencieRepository,
  RideRepository,
  RideUpdateInterface,
  RideQueryInterface,
} from "@app/repositories";
import { EVENTS, State, Position, EventsInterface } from "@shared/events";
import { CANCELATION } from "../constants";
import { StateService } from "../state.service";
import { NotInRideException, RideNotFoundException } from "../exceptions";

export class Common implements OnGatewayInit<Server>, OnGatewayConnection {
  public role!: UserRoles;

  constructor(
    readonly configService: ConfigService,
    readonly socketService: SocketService<EventsInterface>,
    readonly rideRepository: RideRepository,
    readonly pendencieRepository: PendencieRepository,
    readonly sessionService: AuthService,
    readonly stateService: StateService,
    readonly cacheService: CacheService,
    readonly logger: PinoLogger,
  ) {
    logger.setContext(Common.name);
  }

  afterInit(server: Server) {
    auth.server(server, async ({ socket, data: { token, p2p } }) => {
      const session = await this.sessionService.verify(
        token,
        socket.handshake.address,
      );

      const hasPermission = this.sessionService.hasPermission(session, [
        this.role,
      ]);

      if (!hasPermission) {
        throw new ForbiddenException();
      }

      const { _id, pid, averageEvaluation } = session.user;

      socket.data = await this.stateService.setConnectionData(pid, {
        _id,
        pid,
        mode: this.role,
        p2p,
        rate: averageEvaluation,
        socketId: socket.id,
        rides: [],
      });

      return true;
    });

    if (process.env.NODE_ENV === "production") {
      unique(server, {
        storage: new storageAdapters.IORedis(this.cacheService.redis),
        onError: (local, error, socket) =>
          this.logger.error(error.message, { local, socketId: socket.id }),
      });
    }
  }

  handleConnection(socket: Socket) {
    socket.auth.catch(this.logger.warn);
  }

  positionEventHandler(position: Position, socket: Socket) {
    this.dispachToObervers(EVENTS.POSITION, socket, position);
  }

  @SubscribeMessage(EVENTS.STATE)
  stateEventHandler(
    @MessageBody() state: State,
    @ConnectedSocket() socket: Socket,
  ) {
    socket.state = state.state;
    this.dispachToObervers(EVENTS.STATE, socket, state);
  }

  dispachToObervers<K extends keyof EventsInterface>(
    event: keyof EventsInterface,
    client: Socket,
    data: EventsInterface[K],
    considerP2P = true,
  ) {
    const { observers, p2p: selfP2P } = client.data;

    data = this.signObservableEvent(data, client);

    observers.forEach((observer) => {
      if (considerP2P && selfP2P && observer.p2p) {
        return;
      }

      this.socketService.emit(observer.socketId, event, data);
    });
  }

  signObservableEvent<T = any>(packet: T, socket: Socket): T {
    return { ...packet, pid: socket.data.pid };
  }

  public updateRide(query: RideQueryInterface, data: RideUpdateInterface) {
    return retryUnderHood(() => this.rideRepository.update(query, data));
  }

  public createPendencie({
    ride,
    issuer,
    affected,
  }: {
    ride: IRide["_id"];
    issuer: IUser["_id"];
    affected: IUser["_id"];
  }) {
    return retryUnderHood(
      () =>
        this.pendencieRepository.create({
          issuer,
          affected,
          amount: CANCELATION.FARE,
          ride,
        }),
      3,
      500,
    );
  }

  async getRide(query: RideQueryInterface): Promise<IRide> {
    const ride = await this.rideRepository.get(query);

    if (!ride) {
      throw new RideNotFoundException();
    }

    return ride;
  }

  checkIfInRide(ride: IRide, _id: IUser["_id"]) {
    if (ride.voyager._id !== _id && ride.driver?._id !== _id) {
      throw new NotInRideException(ride.pid, _id);
    }
  }

  isSafeCancel(acceptTimestamp: number, now: number) {
    const cancelationSafeTime = this.configService.get(
      "OFFER.SAFE_CANCELATION_WINDOW",
    ) as number;
    return acceptTimestamp + cancelationSafeTime > now;
  }
}
