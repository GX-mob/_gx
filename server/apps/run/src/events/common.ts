import {
  MessageBody,
  SubscribeMessage,
  OnGatewayInit,
  OnGatewayConnection,
  OnGatewayDisconnect,
  ConnectedSocket,
} from "@nestjs/websockets";
import { EventEmitter } from "eventemitter3";
import { SocketService, SocketModule } from "@app/socket";
import { Position } from "../schemas/events/position";
import { Connection } from "../schemas/common/connection";
import { Server, Socket } from "socket.io";
import { Module } from "@nestjs/common";
import { USERS_ROLES } from "@app/database";
import { CacheModule, CacheService } from "@app/cache";
import { SessionModule, SessionService } from "@app/session";
import { CACHE_NAMESPACES, CACHE_TTL } from "../constants";
import { auth } from "extensor";

declare module "socket.io" {
  interface Socket {
    connection: Connection;
  }
}

@Module({ imports: [CacheModule, SessionModule, SocketModule] })
export class Common
  extends EventEmitter
  implements
    OnGatewayInit<Server>,
    OnGatewayConnection<Socket>,
    OnGatewayDisconnect<Socket> {
  public role!: USERS_ROLES;

  constructor(
    readonly socketService: SocketService,
    readonly cacheService: CacheService,
    readonly sessionService: SessionService,
  ) {
    super();
  }

  afterInit(server: Server) {
    auth.server(server, async ({ socket, data: { token, p2p } }) => {
      const session = await this.sessionService.verify(
        token,
        socket.handshake.address,
      );

      const hasPermission = this.sessionService.hasPermission(session, [
        this.role,
      ]);

      if (!hasPermission) {
        throw new Error("unauthorized");
      }

      const { pid, averageEvaluation } = session.user;
      const previous = await this.getConnection(pid);

      const connection: Connection = {
        _id: session.user._id,
        pid,
        mode: this.role,
        p2p,
        observers: previous.observers || [],
        rate: averageEvaluation,
        socketId: socket.id,
      };

      await this.setConnection(pid, connection);

      (socket as any).connection = connection;

      return true;
    });
  }

  handleConnection(socket: Socket) {
    console.log("client connected", socket.id);
  }

  handleDisconnect(socket: Socket) {
    console.log("client disconected", socket.id);
  }

  @SubscribeMessage("position")
  async positionEventHandler(
    @MessageBody() position: Position,
    @ConnectedSocket() client: Socket,
  ) {
    this.emit("position", position);
    this.dispachToObervers(
      "position",
      client,
      this.signObservableEvent(position, client),
    );
  }

  dispachToObervers<T = any>(
    event: string,
    client: Socket,
    data: T,
    considerP2P = true,
  ) {
    const { observers } = client.connection;
    for (let i = 0; i < observers.length; ++i) {
      if (considerP2P && observers[i].p2p) {
        continue;
      }
      this.socketService.emit(observers[i].socketId, event, data);
    }
  }

  signObservableEvent<T = any>(packet: T, client: Socket): T {
    return { ...packet, pid: client.connection.pid };
  }

  /**
   * Get connection data
   * @param id Socket ID or User public ID
   */
  public getConnection(id: string): Promise<Connection> {
    return this.cacheService.get(CACHE_NAMESPACES.CONNECTIONS, id);
  }

  /**
   * Set connection data
   * @param pid User public ID
   * @param data
   */
  public async setConnection(
    pid: string,
    data: Connection,
  ): Promise<Connection> {
    const previousData = await this.cacheService.get(
      CACHE_NAMESPACES.CONNECTIONS,
      pid,
    );

    const newData = { ...previousData, ...data };

    await this.cacheService.set(CACHE_NAMESPACES.CONNECTIONS, pid, newData, {
      link: ["socketId"],
      ex: CACHE_TTL.CONNECTIONS,
    });
    return newData;
  }
}
