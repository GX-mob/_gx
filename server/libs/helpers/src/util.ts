/**
 * GX - Corridas
 * Copyright (C) 2020  Fernando Costa
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import SecurePassword from "secure-password";
export { getClientIp } from "request-ip";
import logger from "./logger";
import { UnprocessableEntityException } from "@nestjs/common";
import { Logger } from "pino";
import validator from "validator";

export const securePassword = new SecurePassword();

export const handleRejectionByUnderHood = (
  promise: Promise<unknown>,
  loggerInstance: Logger = logger,
) => {
  promise.catch((error) => loggerInstance.error(error));
};

/**
 * General regexes
 */
export const passwordRegex = /^(?=.*\\d)(?=.*[a-z]).{5,}$/;

export type PhoneContactObject = {
  cc: string;
  number: string;
};

export type ContactResultObject = {
  value: string;
  field: "emails" | "phones";
};

/**
 * Validates a contact
 *
 * @param {stirng} value
 * @throws UnprocessableEntityException("invalid-contact")
 * @return {object} { value: string, type: "email" | "phone" }
 */
export function isValidContact(value: string): ContactResultObject {
  const isMobilePhone = validator.isMobilePhone(value);

  if (!isMobilePhone && !validator.isEmail(value)) {
    throw new UnprocessableEntityException("invalid-contact");
  }

  return {
    field: isMobilePhone ? "phones" : "emails",
    value,
  };
}

/**
 * Creates a password buffer hash
 * @param password
 */
export function hashPassword(password: string | Buffer): Promise<Buffer> {
  return securePassword.hash(
    typeof password === "string" ? Buffer.from(password) : password,
  );
}

/**
 * Compare password
 * @param value Plain text value
 * @param to Hashed password
 * @returns {Promise<Buffer | boolean>} Buffer if it was necessary to rehash or the verification result
 */
export async function assertPassword(
  value: string,
  to: Buffer,
): Promise<Buffer | boolean> {
  const passwordValue = Buffer.from(value);
  const result = await securePassword.verify(passwordValue, to);

  switch (result) {
    case SecurePassword.VALID:
      return true;
    case SecurePassword.VALID_NEEDS_REHASH:
      return hashPassword(passwordValue);
    default:
      return false;
  }
}
/**
 * Retries the given function until it succeeds given a number of retries and an interval between them. They are set
 * by default to retry 5 times with 1sec in between. There's also a flag to make the cooldown time exponential
 * @author Daniel IÃ±igo <danielinigobanos@gmail.com>
 * @author Thijs Koerselman <gitlab.com/0x80>
 * @param {Function} fn - Returns a promise
 * @param {Number} retriesLeft - Number of retries. If -1 will keep retrying
 * @param {Number} interval - Millis between retries. If exponential set to true will be doubled each retry
 * @param {Boolean} exponential - Flag for exponential back-off mode
 * @return {Promise<*>}
 */
export async function retry<T>(
  fn: () => Promise<T>,
  retriesLeft: number = 3,
  interval: number = 1000,
  exponential: boolean = false,
): Promise<T> {
  try {
    const val = await fn();
    return val;
  } catch (error) {
    if (retriesLeft) {
      await new Promise((r) => setTimeout(r, interval));
      return retry(
        fn,
        retriesLeft - 1,
        exponential ? interval * 2 : interval,
        exponential,
      );
    } else
      throw new Error(
        `Max retries reached for function ${fn.name}, last error: ${error.message}`,
      );
  }
}
