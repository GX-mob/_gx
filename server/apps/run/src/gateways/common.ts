import {
  MessageBody,
  SubscribeMessage,
  OnGatewayInit,
  OnGatewayConnection,
  OnGatewayDisconnect,
  ConnectedSocket,
} from "@nestjs/websockets";
import { SocketService, SocketModule } from "@app/socket";
import { Server, Socket } from "socket.io";
import { Module } from "@nestjs/common";
import { USERS_ROLES } from "@app/database";
import { CacheModule, CacheService } from "@app/cache";
import { SessionModule, SessionService } from "@app/session";
import { auth } from "extensor";
import { OffersState } from "../states/offers.state";
import { DriversState } from "../states/drivers.state";
import { ConnectionDataService } from "../conn-data.service";
import { DataModule } from "@app/data";
import { EVENTS, State, Position, Connection } from "../events";
declare module "socket.io" {
  interface Socket {
    connection: Connection;
  }
}

@Module({
  imports: [CacheModule, DataModule, SessionModule, SocketModule],
  providers: [DriversState, OffersState, ConnectionDataService],
})
export class Common
  implements
    OnGatewayInit<Server>,
    OnGatewayConnection<Socket>,
    OnGatewayDisconnect<Socket> {
  public role!: USERS_ROLES;

  constructor(
    readonly socketService: SocketService,
    readonly cacheService: CacheService,
    readonly sessionService: SessionService,
    readonly driversState: DriversState,
    readonly offersState: OffersState,
    readonly connectionData: ConnectionDataService,
  ) {}

  afterInit(server: Server) {
    auth.server(server, async ({ socket, data: { token, p2p } }) => {
      const session = await this.sessionService.verify(
        token,
        socket.handshake.address,
      );

      const hasPermission = this.sessionService.hasPermission(session, [
        this.role,
      ]);

      if (!hasPermission) {
        throw new Error("unauthorized");
      }

      const { pid, averageEvaluation } = session.user;
      const previousData = await this.connectionData.get(pid);

      const connection: Connection = {
        _id: session.user._id,
        pid,
        mode: this.role,
        p2p,
        observers: previousData.observers || [],
        rate: averageEvaluation,
        socketId: socket.id,
      };

      await this.connectionData.set(pid, connection);

      socket.connection = connection;

      return true;
    });
  }

  handleConnection(socket: Socket) {
    console.log("client connected", socket.id);
  }

  handleDisconnect(socket: Socket) {
    console.log("client disconected", socket.id);
  }

  positionEventHandler(position: Position, client: Socket) {
    this.dispachToObervers("position", client, position);
  }

  @SubscribeMessage(EVENTS.STATE)
  stateEventHandler(
    @MessageBody() state: State,
    @ConnectedSocket() client: Socket,
  ) {
    // client.connection.state = state.state;
    this.dispachToObervers("state", client, state);
  }

  dispachToObervers<T = any>(
    event: string,
    client: Socket,
    data: T,
    considerP2P = true,
  ) {
    const { observers } = client.connection;

    data = this.signObservableEvent(data, client);

    observers.forEach((observer) => {
      if (considerP2P && observer.p2p) {
        return;
      }

      this.socketService.emit(observer.socketId, event, data);
    });
  }

  signObservableEvent<T = any>(packet: T, client: Socket): T {
    return { ...packet, pid: client.connection.pid };
  }
}
