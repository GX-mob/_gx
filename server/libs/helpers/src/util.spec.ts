/**
 * Helper: util
 *
 * @group unit/helpers/util
 */
import * as util from "./util";
import faker from "faker";
import SecurePassword from "secure-password";
import { UnprocessableEntityException } from "@nestjs/common";

const defer = (): {
  promise: Promise<unknown>;
  resolve: typeof Promise.resolve;
  reject: typeof Promise.reject;
} => {
  const defered: any = {};

  defered.promise = new Promise((resolve, reject) => {
    defered.resolve = resolve;
    defered.reject = reject;
  });

  return defered;
};

describe("Helper: util", () => {
  it("handleRejectionByUnderHood", async () => {
    const error = new Error("error");

    const defered = defer();
    const logger = { error: jest.fn() };

    util.handleRejectionByUnderHood(defered.promise, logger as any);

    defered.reject(error);

    // wait catch callback call
    await new Promise((resolve) => setTimeout(resolve, 10));

    expect(logger.error.mock.calls[0][0]).toStrictEqual(error);
  });

  describe("isValidContact", () => {
    it("validate contacts", () => {
      const email = faker.internet.email();
      const mobilePhone = "+5582988888888";

      expect(util.isValidContact(email)).toStrictEqual({
        field: "emails",
        value: email,
      });
      expect(util.isValidContact(mobilePhone)).toStrictEqual({
        field: "phones",
        value: mobilePhone,
      });
    });

    it('should throw UnprocessableEntityException("invalid-contact")', () => {
      const invalidMobilePhone = "+5588888";
      const invalidEmail = "Ã¢d@.com";

      expect(() => util.isValidContact(invalidMobilePhone)).toThrow(
        new UnprocessableEntityException("invalid-contact"),
      );

      expect(() => util.isValidContact(invalidEmail)).toThrow(
        new UnprocessableEntityException("invalid-contact"),
      );
    });
  });

  describe("hashPassword", () => {
    it("hash string", async () => {
      expect((await util.hashPassword("foo")) instanceof Buffer).toBeTruthy();
    });

    it("hash buffer", async () => {
      expect(
        (await util.hashPassword(Buffer.from("foo"))) instanceof Buffer,
      ).toBeTruthy();
    });
  });

  describe("assertPassword", () => {
    it("verify a valid password", async () => {
      const plainPassword = faker.internet.password();
      const hashedPassword = await util.hashPassword(plainPassword);

      expect(
        await util.assertPassword(plainPassword, hashedPassword),
      ).toBeTruthy();
    });

    it("verify a invalid password", async () => {
      const plainPassword = faker.internet.password();
      const diferentPassword = faker.internet.password();
      const hashedPassword = await util.hashPassword(plainPassword);

      expect(
        await util.assertPassword(diferentPassword, hashedPassword),
      ).toBeFalsy();
    });

    it("return password hehashed", async () => {
      const weakSecurePasswordInstance = new SecurePassword({
        memlimit: SecurePassword.MEMLIMIT_DEFAULT / 2,
        opslimit: SecurePassword.OPSLIMIT_DEFAULT / 2,
      });

      const plainPassword = faker.internet.password();
      const weakHash = await weakSecurePasswordInstance.hash(
        Buffer.from(plainPassword),
      );

      expect(
        (await util.assertPassword(plainPassword, weakHash)) instanceof Buffer,
      ).toBeTruthy();
    }, 100000);
  });

  describe("retry", () => {
    it("should resolve on first try", async () => {
      const resolvedValue = "foo";
      const promise: any = jest.fn();

      promise.mockResolvedValue(resolvedValue);

      await expect(util.retry(promise)).resolves.toBe(resolvedValue);
    });

    it("should resolve on second try", async () => {
      const resolvedValue = "foo";
      const promise: any = jest.fn();

      promise.mockRejectedValueOnce(new Error("any"));
      promise.mockResolvedValueOnce(resolvedValue);

      await expect(util.retry(promise)).resolves.toBe(resolvedValue);
    });

    it("should hit max retries", async () => {
      const promise: any = jest.fn();
      const maxRetrys = 2;
      const errorMsg = "any3";

      promise.mockRejectedValueOnce(new Error(errorMsg));
      promise.mockRejectedValueOnce(new Error(errorMsg));
      promise.mockRejectedValueOnce(new Error(errorMsg));

      await expect(util.retry(promise, maxRetrys)).rejects.toStrictEqual(
        new Error(errorMsg),
      );
    });
  });

  describe("hasProp", () => {
    const obj = { foo: "bar" };
    const unsecureObj = { foo: "bar", hasOwnProperty: (prop: string) => true };

    it("should check over object", () => {
      expect(util.hasProp(obj, "foo")).toBeTruthy();
      expect(util.hasProp(obj, "bar")).toBeFalsy();
    });

    it("should check over unsecure object", () => {
      // direct call, false positive
      expect(unsecureObj.hasOwnProperty("not-have")).toBeTruthy();

      expect(util.hasProp(unsecureObj, "not-have")).toBeFalsy();
      expect(util.hasProp(unsecureObj, "hasOwnProperty")).toBeTruthy();
      expect(util.hasProp(unsecureObj, "foo")).toBeTruthy();
      expect(util.hasProp(unsecureObj, "bar")).toBeFalsy();
    });
  });

  describe("hideEmail", () => {
    const email = "valid@email.com";

    it("should hide", () => {
      expect(util.hideEmail(email)).toBe("val**@email.com");
    });

    it("should change default visible chars", () => {
      expect(util.hideEmail(email, 2)).toBe("va***@email.com");
    });

    it("should hide provider", () => {
      expect(util.hideEmail(email, 3, 2)).toBe("val**@em*******");
    });
  });

  describe("decimalAdjust", () => {
    const decimal1 = 1.2345123;
    const decimal2 = 2.6246456;
    const decimal3 = 3.3425212;
    const decimal4 = 4.1645212;

    it("round adjust", () => {
      expect(util.decimalAdjust(decimal1, -3)).toMatchSnapshot();
      expect(util.decimalAdjust(decimal2, -2)).toMatchSnapshot();
      expect(util.decimalAdjust(decimal3, -1)).toMatchSnapshot();
      expect(util.decimalAdjust(decimal4)).toMatchSnapshot();
    });

    it("floor adjust", () => {
      expect(util.decimalAdjust(decimal1, -3, "floor")).toMatchSnapshot();
      expect(util.decimalAdjust(decimal2, -2, "floor")).toMatchSnapshot();
      expect(util.decimalAdjust(decimal3, -1, "floor")).toMatchSnapshot();
      expect(util.decimalAdjust(decimal4, 0, "floor")).toMatchSnapshot();
    });

    it("ceil adjust", () => {
      expect(util.decimalAdjust(decimal1, -3, "ceil")).toMatchSnapshot();
      expect(util.decimalAdjust(decimal2, -2, "ceil")).toMatchSnapshot();
      expect(util.decimalAdjust(decimal3, -1, "ceil")).toMatchSnapshot();
      expect(util.decimalAdjust(decimal4, 0, "ceil")).toMatchSnapshot();
    });
  });
});
