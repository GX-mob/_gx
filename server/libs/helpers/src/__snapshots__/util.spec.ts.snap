// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helper: util decimalAdjust ceil adjust 1`] = `1.235`;

exports[`Helper: util decimalAdjust ceil adjust 2`] = `2.63`;

exports[`Helper: util decimalAdjust ceil adjust 3`] = `3.4`;

exports[`Helper: util decimalAdjust ceil adjust 4`] = `5`;

exports[`Helper: util decimalAdjust floor adjust 1`] = `1.234`;

exports[`Helper: util decimalAdjust floor adjust 2`] = `2.62`;

exports[`Helper: util decimalAdjust floor adjust 3`] = `3.3`;

exports[`Helper: util decimalAdjust floor adjust 4`] = `4`;

exports[`Helper: util decimalAdjust round adjust 1`] = `1.235`;

exports[`Helper: util decimalAdjust round adjust 2`] = `2.62`;

exports[`Helper: util decimalAdjust round adjust 3`] = `3.3`;

exports[`Helper: util decimalAdjust round adjust 4`] = `4`;
